openslides/utils/cache.py

```diff
@@ -1,506 +1,424 @@
-            id)
+        Returns one element as full data.
 
-    def exists_for_collection(self, collection_string: str) -> bool:
-        """
-        Returns True if the cache for the collection exists, else False.
-        """
-        redis = get_redis_connection()
-        return redis.exists(self.get_cache_key(collection_string))
+        If the cache is empty, it is created.
 
-    def get_data(self, collection_string: str) -> List[Dict[str, Any]]:
+        Returns None if the element does not exist.
         """
-        Returns all data for the collection.
-        """
-        redis = get_redis_connection()
-        return [json.loads(element.decode()) for element in redis.hvals(self.get_cache_key(collection_string))]
+        if not await self.exists_full_data():
+            await self.build_full_data()
 
-    def get_element(self, collection_string: str, id: int) -> Dict[str, Any]:
-        """
-        Returns one element from the collection.
+        element = await self.cache_provider.get_element(get_element_id(collection_string, id))
 
-        Raises model.DoesNotExist if the element is not in the cache.
-        """
-        redis = get_redis_connection()
-        element = redis.hget(self.get_cache_key(collection_string), id)
         if element is None:
-            from .collection import get_model_from_collection_string
-            model = get_model_from_collection_string(collection_string)
-            raise model.DoesNotExist(collection_string, id)
+            return None
         return json.loads(element.decode())
 
-    def get_cache_key(self, collection_string: str) -> str:
+    async def exists_restricted_data(self, user: Optional['CollectionElement']) -> bool:
+        """
+        Returns True, if the restricted_data exists for the user.
+        """
+        if not self.use_restricted_data_cache:
+            return False
+
+        return await self.cache_provider.data_exists(get_user_id(user))
+
+    async def del_user(self, user: Optional['CollectionElement']) -> None:
+        """
+        Removes one user from the resticted_data_cache.
+        """
+        await self.cache_provider.del_restricted_data(get_user_id(user))
+
+    async def update_restricted_data(
+            self, user: Optional['CollectionElement']) -> None:
+        """
+        Updates the restricted data for an user from the full_data_cache.
+        """
+        # TODO: When elements are changed at the same time then this method run
+        #       this could make the cache invalid.
+        #       This could be fixed when get_full_data would be used with a
+        #       max change_id.
+        if not self.use_restricted_data_cache:
+            # If the restricted_data_cache is not used, there is nothing to do
+            return
+
+        # Try to write a special key.
+        # If this succeeds, there is noone else currently updating the cache.
+        # TODO: Make a timeout. Else this could block forever
```

Comment:
Maybe you have to use a asyncio.Lock https://docs.python.org/3/library/asyncio-sync.html#lock

Reply:
Let's talk about this on Wednesday.

Reply:
The problem is, that we need a lock, that is also realized in other daphne instances.